packages/eslint-plugin/src/rules/unified-signatures.ts

```diff
@@ -310,6 +314,12 @@ export default createRule<Options, MessageIds>({
       const shorter = sig1.length < sig2.length ? sig1 : sig2;
       const shorterSig = sig1.length < sig2.length ? a : b;
 
+      // If one signature has explicit this type and another doesn't, they can't
+      // be unified.
+      if (isThisParam(sig1[0]) !== isThisParam(sig2[0])) {
```

Comment:
Question -

```suggestion
      if (isThisParam(sig1[0]) || isThisParam(sig2[0])) {
```

I was kind of thinking we would also want to exempt cases where they both include `this`, since `this: Foo | void` isn't equivalent to the overload form ([playground](https://typescript-eslint.io/play/#ts=5.7.2&fileType=.ts&code=GYVwdgxgLglg9mABHAbgUwE4Bs4EMAma%2BAKgBYwDOAanDPgBRTkUBciKt%2BAlGx3QNwAoUJFgJk6bHkIlmNOo2ZsKUDDDABzHu05CR0eElSYcBImUryGXRAG9EAX0HDwB8eEMB1GEyuLKyqrqGogAPjp02nz4do7OEAgqyIgAvLHGUmaylpwANIgeCN6%2BnHGCGaYyFtSc9FxCcAB0FdLmcrX15ZKVbTl0jRC4WFj0GGhQIBhgNQp1qQB8iADkOFhLXJ2ChWDFpH6dTdu7%2B0JHPnucA0MjYxNTMwxzKYsrcGsbQi6ihoi3k9N%2BCAAIzYT0W0SipVsgkQv3G-0QwLqQicgjQAA8AA5wDBQWIOfhAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6WJygM0sQBNaySgHMmAQ3yxoKdFADuY6E0jgwAXxBqgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false)). Is that not the case?

Reply:
Hmm but using different logic, we could also exempt cases where some but not all signatures have `this: void` without breaking the above example

Reply:
@kirkwaiblinger btw does `returnsVoid` (especially calling it with a function argument) do something tricky in your example that I'm not aware of?  It seems to me instead of

```ts
overloadedThisVoid.call(returnsVoid(() => 'lol'));
```

you could just do
```ts
overloadedThisVoid.call(1 as any as void)
```

Which triggers the same TypeScript error.

Reply:
Hmm, yeah, I guess it would come down to special-casing around `void`, since, in addition to your example, which I agree with, we'd also want to consider:
```ts
function foo(this: Foo, x: number)
function foo(this: Bar, x: number)
```
can still be
```ts
function foo(this: Foo | Bar, x: number)
```

Wonder if special-casing around `void` is even relevant for non-`this` parameters too?

```ts
function foo(x: void);
function foo(x: number);
```

Would have to play around to figure that one out.

Reply:
Yes the `Foo | Bar` case was another one I had in mind.

`void` on non-`this` parameters doesn't seem like it would cause any problems with unifying signatures, but interestingly, it does allow trailing parameters to be omitted:

```ts
function foo(x: void): void;
function foo(x: number): void;
function foo() {}

foo()
foo((() => {})())
foo(1)

function bar(x: void | number) {}

bar()
bar((() => {})())
bar(1)
```

Reply:
> @kirkwaiblinger btw does `returnsVoid` (especially calling it with a function argument) do something tricky in your example that I'm not aware of? It seems to me instead of
> 
> ```ts
> overloadedThisVoid.call(returnsVoid(() => 'lol'));
> ```
> 
> you could just do
> 
> ```ts
> overloadedThisVoid.call((() => {})())
> ```
> 
> Or more explicitly
> 
> ```ts
> overloadedThisVoid.call(1 as any as void)
> ```
> 
> Both of which trigger the same TypeScript error.

Sort of - the point of those in the issue was to demonstrate how you could sneak _any_ runtime value you like into `void` without any TS errors or type assertions. 

So, 
```ts
overloadedThisVoid.call((() => {})())
```
doesn't demonstrate that, since the runtime value actually is `undefined`, and 
```ts
overloadedThisVoid.call(1 as any as void)
```
doesn't demonstrate that, due to the type assertions.

It isn't as pertinent to this discussion anymore, yeah, since we're now just talking about what happens when the `void` type is used, not whether using `void` type is unsafe. But that's just the examples that have been on my mind 🙂

Reply:
It appears that TS picks the `this` type from the last overload, lol.

```
function foo(this: string): void;
function foo(this: number): void;
function foo() {}

foo.call(1)
foo.call('a') // Argument of type 'string' is not assignable to parameter of type 'number'.(2345)

function bar(this: number): void;
function bar(this: string | boolean): void;
function bar() {}

bar.call(1) // Argument of type 'number' is not assignable to parameter of type 'string'.(2345)
bar.call('a')
bar.call(true)
```

Reply:
ah right, the unsoundness of TS function typing rears its head again

Reply:
> It appears that TS picks the `this` type from the last overload when it comes to `.call`, lol.

Huh. 👀 . Well - that the `.bind()`/`.call()`/`.apply()` types are going over my head at this point 😆

Reply:
> Do you know why TS doesn't treat embedded void expressions as an error the way `no-confusing-void-expression` does? Would be curious to know the explicit rationale from the TS team.

Not off the top of my head, but I'm sure you'll get (or more likely be pointed to) a thorough answer if you ask in the TS discord.